appengine/genetics/js/cage.js

```diff
@@ -0,0 +1,725 @@
+    }
+  }
+  // Failed to retrieve a return value.
+  Genetics.Cage.death(mouse);
+  return ['FAILURE'];
+};
+
+/**
+ * Returns an interpreter with player code from the mouse for the given
+ * mouseFunction with a call appended.
+ * @param {!Genetics.Mouse} mouse The mouse to get the code from.
+ * @param {string} mouseFunction The function call to append to the code in the
+ * interpreter.
+ * @param {Genetics.Mouse=} opt_suitor The mouse passed as a parameter to the
+ * function (for mateAnswer function call).
+ * @return {!Interpreter} Interpreter set up for executing mouse function call.
+ */
+Genetics.Cage.getInterpreter = function(mouse, mouseFunction, opt_suitor) {
+  var playerId;
+  switch (mouseFunction) {
+    case 'pickFight':
+      playerId = mouse.pickFightOwner;
+      break;
+    case 'chooseMate':
+      playerId = mouse.chooseMateOwner;
+      break;
+    case 'mateAnswer':
+      playerId = mouse.mateAnswerOwner;
+      break;
+  }
+  var code = Genetics.Cage.players[playerId][1];
+  if (goog.isFunction(code)) {
+    code = code();
+  } else if (!goog.isString(code)) {
+    var player = Genetics.Cage.players[playerId][0];
+    throw 'Player ' + player + ' has invalid code: ' + code;
+  }
+  var interpreter = new Interpreter(code,
+      goog.partial(Genetics.Cage.initInterpreter, mouse, opt_suitor));
+  // Overwrite other function calls and call function we need return value of.
+  switch (mouseFunction) {
+    case 'pickFight':
+      interpreter.appendCode('chooseMate = null;');
+      interpreter.appendCode('mateAnswer = null;');
+      interpreter.appendCode('pickFight();');
+      break;
+    case 'chooseMate':
+      interpreter.appendCode('pickFight = null;');
+      interpreter.appendCode('mateAnswer = null;');
+      interpreter.appendCode('chooseMate();');
+      break;
+    case 'mateAnswer':
+      interpreter.appendCode('pickFight = null;');
+      interpreter.appendCode('chooseMate = null;');
+      interpreter.appendCode('mateAnswer(getSuitor());');
+      break;
+  }
+  return interpreter;
+};
+
+/**
+ * Inject the Genetics API into a JavaScript interpreter.
+ * @param {!Genetics.Mouse} mouse The mouse that is running the function.
+ * @param {!Genetics.Mouse|undefined} suitor The mouse passed as a parameter to
+ * the function (for mateAnswer function call).
+ * @param {!Interpreter} interpreter The JS interpreter.
+ * @param {!Object} scope Global scope.
+ */
+Genetics.Cage.initInterpreter = function(mouse, suitor, interpreter,
+    scope) {
+  var pseudoMe = interpreter.UNDEFINED;
+  var pseudoSuitor = interpreter.ARRAY;
+  var pseudoAliveMice = interpreter.createObject(interpreter.ARRAY);
+  var aliveMice = Genetics.Cage.aliveMice_;
+  for (var i = 0; i < aliveMice.length; i++) {
+    var aliveMouse = aliveMice[i];
+    // Create a clone of alive mouse with string keys so that keys wont be
+    // renamed when compressed.
+    var clonedMouse = {
+      'pickFightOwner': aliveMice.pickFightOwner,
```

Comment:
Done. It wasn't caught because current AI don't use function owners and if they did it wouldn't error but just consider the function owner to be undefined.